elm/Swole/Components/MovementInput.elm

```diff
@@ -0,0 +1,112 @@
+module Swole.Components.MovementInput exposing
+    ( Msg
+    , Model
+    , view
+    , update
+    )
+
+import Html exposing (Html, program, div, input, text, button)
+import Html.Attributes exposing (placeholder, type_, value, id)
+import Html.Events exposing (onClick, onInput)
+import Task
+import Dom exposing (focus)
+import List.Extra as List
+
+import Helpers.List as List
+import Helpers.Events exposing (onDeleteWhen)
+import Swole.Types.Movement exposing (Movement)
+
+type alias Movements = List (Int, Movement)
+type alias Model = Movements
+
+type Msg
+    = NoOp
+    | MovementChanged (Int, Movement)
+    | AddMovement
+    | DeleteMovement Int
+
+view : Movements -> Html Msg
+view movements =
+    div []
+        (movementFields movements)
+
+update : Msg -> Movements -> (Movements, Cmd Msg)
+update msg movements =
+    case msg of
+        NoOp ->
+            (movements, Cmd.none)
+
+        MovementChanged (idx, m) ->
+            let
+                (newIdx, newMovements) = splitMovements (idx, m)
+                (before, after) = movementsAround idx movements
+                allMovements = before ++ newMovements ++ after
+            in
+                (List.enumerated allMovements, updateFocus newIdx)
+
+        AddMovement ->
+            let
+                rawMovements = List.map Tuple.second movements
+                idx = List.length rawMovements
+            in
+                (List.enumerated <| rawMovements ++ [""], updateFocus idx)
+
+        DeleteMovement idx ->
+            let
+                newMovements
+                    = movements
+                    |> List.map Tuple.second
+                    |> List.removeAt idx
+
+                newIdx = max 0 (idx - 1)
+            in
+               (List.enumerated newMovements, updateFocus newIdx)
+
+movementsAround : Int -> Movements -> (List Movement, List Movement)
+movementsAround idx movements
+    = movements
+    |> List.map Tuple.second
```

Comment:
Having looked at this file, there are a significant number of functions that interact with the list of pairs. I'm wondering if there's a more concrete, custom type that could be built, with a handful of functions specific to that type, that better convey the intent here. Specifically, updating movements and deleting movements have a decent amount of complexity.

Reply:
Yes, for sure. This is also going to be a common pattern, so I'm very interested in finding a more reasonable solution to this.

Reply:
I think for now, I'd build up a type to capture this concept, and think about generalization after it crops up again?

Reply:
It technically already did crop up again, this is the exact same pattern I'm using inside Main for handling multiple sets. I'm going to move forward with this and open up a new PR trying to move to `SelectList`.